### PR TITLE
feat: show multiple diagram instances on a grid in `editor`

### LINKED ISSUE #1288: Name collisions in SVG `Image` import
**Describe the bug**

When there are more than one diagram on the page, imported images might not show up. This is usually caused by name collisions among the imported SVGs. If an imported SVG defines a `<linearGradient id="foo">` and refer to `foo` by `url(#foo)` somewhere else, there must be only one such reference in the entire DOM tree. 

**To Reproduce**

1. Go to https://361060b0.penrose-72l.pages.dev/try/
2. Click on the "Walk on Spheres" example
3. Select `Grid`
4. See the following output

image

**Expected behavior**

Spheres should be shaded by the imported image:

Walk on Spheres - Laplace Estimator

**Additional context**
This is a known problem with SVG (e.g. https://github.com/gregberge/svgr/issues/150) and there are some SVG cleaning tools like [svgr](https://github.com/gregberge/svgr). It would be great to experiment with some of them.

### PR DESCRIPTION
**Pull request by wodeni**
# Description

This PR introduces a new `Grid` panel to `editor`, which show multiple diagram instances at once. The number of instances is 10 by default, and configuration through the `option` panel. 

`synthesizer-ui` first introduced a grid of Penrose-generated diagrams in a web app. This PR pulls relevant components (i.e. `Gridbox` and `Grid`) into `components` and uses them in both `editor` and `synthesizer-ui`.

# Implementation strategy and design decisions

* `Grid` contains `Gridbox`, which contains `Simple`. The properties are passed through, with `Gridbox` inheriting all `Simple`'s props and `Grid` having a `gridboxProps` option for each `Gridbox`. 
* `Gridbox` can be `stateful`, which means that it caches the trio it receives from props _and_ caches `state` from `Simple`'s `onFrame` callback.
* Styling: use `styled-components` to style `Grid` and `Gridbox`. The client may provide a theme using `<ThemeProvider>`. 
* `editor`'s state management: the IDE wraps `GridPanel` around `Grid` and maintains a `diagramGridState`. All state updates are done through Recoil. 
* `Image` and `Line` arrowhead id collisions: since SVG has a global namespace, when there are multiple diagrams on the page with duplicated defs, they disappear. This PR introduces an util function `makeIDsUnique`, which changes all definitions and corresponding references of [IRI references](https://dev.w3.org/SVG/profiles/1.2T/publish/diff/linking.html) in the SVG such that they are unique. (Resolves #1288)

# Examples with steps to reproduce them

See preview deployment:
grid

# Checklist

- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new ESLint warnings
- [x] I have reviewed any generated registry diagram changes

# Open questions

* Caching in `editor`: the grid doesn't cache anything other than a list of `variation`s, which causes re-rendering whenever the grid size is decreased and then increased. No re-rendering happen when the `gridSize` is decreasing because there are no prop changes.

### CODE REVIEW

**Comment by samestep** [comment]
Looks like CI is failing: https://github.com/penrose/penrose/actions/runs/4108063039/jobs/7088302283

```
TypeError: uuid.v4 is not a function
```

Any idea why?

**Comment by wodeni** [comment]
Turns out that bumping `uuid` to 9.0 everywhere fixed the CI. My guess is a lower version is hoisted somehow.

**Review by samestep — COMMENTED** [review]
Still reviewing the code itself, but had a quick comment while playing with it: would it be possible to disable the "replace diagram with variation when you click on it" in `editor`? Or if that would be too much implementation burden for this PR, could we make a GitHub issue to do that after this PR is merged?

**Comment by samestep** [comment]
Is the code for `packages/core/src/renderer/util.ts` copied from somewhere? If so, could you add a comment to the top of that file with a link to where we got it from, and also the copyright notice and license if applicable?

**Comment by samestep** [comment]
Why does `packages/examples/src/walk-on-spheres/walk-on-spheres-ball.svg` change?

**Comment by samestep** [comment]
I'm getting this error in Storybook for Continuous Map under Grid Component:

```
TypeError: this.props.header is not a function
    at http://localhost:6006/src/Grid.tsx:58:26
    at Array.map (<anonymous>)
    at Grid.innerContent (http://localhost:6006/src/Grid.tsx:50:32)
    at Grid.render (http://localhost:6006/src/Grid.tsx:89:21)
    at finishClassComponent (http://localhost:6006/node_modules/.vite-storybook/deps/chunk-4QL7BUP5.js?v=6233ef11:14267:39)
    at updateClassComponent (http://localhost:6006/node_modules/.vite-storybook/deps/chunk-4QL7BUP5.js?v=6233ef11:14232:32)
    at beginWork (http://localhost:6006/node_modules/.vite-storybook/deps/chunk-4QL7BUP5.js?v=6233ef11:15463:22)
    at beginWork$1 (http://localhost:6006/node_modules/.vite-storybook/deps/chunk-4QL7BUP5.js?v=6233ef11:19248:22)
    at performUnitOfWork (http://localhost:6006/node_modules/.vite-storybook/deps/chunk-4QL7BUP5.js?v=6233ef11:18693:20)
    at workLoopSync (http://localhost:6006/node_modules/.vite-storybook/deps/chunk-4QL7BUP5.js?v=6233ef11:18632:13)
```

**Comment by samestep** [comment]
Similarly I get this error in Storybook under GridBox Component, not right away, but as soon as I click on the diagram:

```
TypeError: Cannot read properties of undefined (reading 'map')
    at Gridbox.render (http://localhost:6006/src/Gridbox.tsx:178:41)
    at finishClassComponent (http://localhost:6006/node_modules/.vite-storybook/deps/chunk-4QL7BUP5.js?v=6233ef11:14267:39)
    at updateClassComponent (http://localhost:6006/node_modules/.vite-storybook/deps/chunk-4QL7BUP5.js?v=6233ef11:14232:32)
    at beginWork (http://localhost:6006/node_modules/.vite-storybook/deps/chunk-4QL7BUP5.js?v=6233ef11:15463:22)
    at beginWork$1 (http://localhost:6006/node_modules/.vite-storybook/deps/chunk-4QL7BUP5.js?v=6233ef11:19248:22)
    at performUnitOfWork (http://localhost:6006/node_modules/.vite-storybook/deps/chunk-4QL7BUP5.js?v=6233ef11:18693:20)
    at workLoopSync (http://localhost:6006/node_modules/.vite-storybook/deps/chunk-4QL7BUP5.js?v=6233ef11:18632:13)
    at renderRootSync (http://localhost:6006/node_modules/.vite-storybook/deps/chunk-4QL7BUP5.js?v=6233ef11:18611:15)
    at recoverFromConcurrentError (http://localhost:6006/node_modules/.vite-storybook/deps/chunk-4QL7BUP5.js?v=6233ef11:18233:28)
    at performSyncWorkOnRoot (http://localhost:6006/node_modules/.vite-storybook/deps/chunk-4QL7BUP5.js?v=6233ef11:18375:28)
```

File: packages/components/src/Simple.tsx
Comment (samestep) [inline packages/components/src/Simple.tsx:163]: I'm not super familiar with how all our state management works here; what's the broad idea behind moving this out of `tick` and into `renderCanvas`?

File: packages/editor/src/components/DiagramPanel.tsx
Comment (samestep) [inline packages/editor/src/components/DiagramPanel.tsx:35]: From the diff it looks like this and `pathResolver` just moved from below to up here; any other changes?

**Review by samestep — COMMENTED** [review]
Looks great! Left a few comments here and above. I'm cool to merge this once the questions about `packages/core/src/renderer/util.ts` and Storybook are resolved.

File: packages/core/src/renderer/util.ts
Comment (samestep) [inline packages/core/src/renderer/util.ts:1]: A bit pedantic but I'd replace the `master` reference here with a specific commit:

```suggestion
// Refactored version of `makeIdsUnique` in https://github.com/iconfu/svg-inject/blob/064ac002930deaf96eefb95eaf953c5ef5287992/src/svg-inject.js
```

**Review by samestep — APPROVED** [review]
Awesome, thanks Nimo!

**Pull request merged**
